### DISCUSSION TITLE
Getting error "No such file or directory:  <path-to-conda-env>\\share\\jupyter\\labextensions\\<pkg_name>\\build_log.json

### DISCUSSION THREAD

**Discussion by celticflute** [zuzukin/whl2conda#150] [Q&A]
I successfully built a conda package from a PyPI wheel, and the conda pkg resides currently in the same directory as the wheel.  I have been unsuccessful, however, in installing that .conda package.  The package in question is a JupyterLab extension, which complicates things, I know.  The format of the whl2conda install command I used is:
`whl2conda install -p <conda-env-path> <fully-qualified-conda-pkg-file-path`
However, I am getting the following error:
`ERROR conda.core.link:_execute(950): An error occurred while installing package '<unknown>::jupyterlab_hide_code-4.0.0-py_0'.
Rolling back transaction: done
[Errno 2] No such file or directory: 'C:\\Users\\<user-name>\\_WASP_dev00\\env-JLab-py311\\share\\jupyter\\labextensions\\jupyterlab_hide_code\\build_log.json'
()`
I think there may be two things going on here:
- The .conda package may be expecting to be in some special location to be installed as a JupyterLab extension
- The conda installation being used internally by whl2conda may be the one in the base Miniconda environment, even though I have the project's conda-env activated and the -p with full path was used
I would appreciate it if anyone could give me any insights into this problem.

**Comment by celticflute** [comment]
**ADDITIONAL INFO:**  The general directory structure within which I am working is the following:
```
C:\Users\<username>\
        <project-directory>\
                env-JLab-py311\  # conda environment for the project, produced via conda in base Miniconda with --copy option
                jupyterlab-hide-code__WHL \ # folder holding the PyPI wheel and .conda package file
        miniconda3\  # base environment of Miniconda
```
What I am trying to accomplish is to install the `jupyterlab-hide-code` JLab extension in the `env-JLab-py311` conda environment for the project.  As I said in my initial post, I do wonder if the `conda` in the base environment is being used instead of the one in `env-JLab-py311` project environment, and, if so, how do I force use of the `conda` in the project environment.  (Getting the CLI to use the right `conda` is something I still struggle with.)  Also, I produced the project `conda` environment with the `--copy` option, because I will eventually deploy my JupyterLab project as a web app, and one can run into problems with deployment of symbolic links.

**Comment by analog-cbarber** [comment]
I am confused. Are you saying that you have more than one conda installation on your machine? If so, I don't think that is recommended. In particular, you should not do a `conda install conda` anywhere except the base environment or else strange behavior can happen. There really should only be one conda installation that lives in the base environment. For this reason, you should also not make conda (or anything that depends on it, such as mamba) be a dependency of any package you need to install outside of base.

The `whl2conda install` command is actually quite simple. It is just doing a conda install of the package but since conda does not install dependencies in that case we read the dependencies and install them separately also using `conda install`.

One thing to try is to just do `conda install -p <conda-env-path> <conda-package-file>` and see if you get the same
error.

You can also build your conda package using the `tree` output format, which will just give you a directory tree with the conda package contents. You can see if there is anything strange about the contents such as a mismatch between the contents and the file list. Is `build_log.json` supposed to be part of your package?

**Reply by celticflute** [reply]
Thank you so much, @analog-cbarber.  This gives me a start, but I do have questions (newbie ones, I will admit)

> There really should only be one conda installation that lives in the base environment.

Agreed ... Now I have only one conda installation ... in the base.  `whl2conda` is (only) there, too.

> It [`whl2conda install`] is just doing a conda install of the package but since conda does not install dependencies in that case we read the dependencies and install them separately also using `conda install`.

- How do I "read the dependencies?"
- Also, the `whl2conda` documentation indicates that `whl2conda install` installs the package _with dependencies_.
- Would `conda install` really not install dependencies?

> One thing to try is to just do `conda install -p <conda-env-path> <conda-package-file>` and see if you get the same
> error.

I tried the above `conda install` as you suggested, but I do get the same error.

> You can also build your conda package using the `tree` output format ...

Build the package with the "`tree` output format" of what command?

> Is `build_log.json` supposed to be part of your package?

Yes, I believe so, and it _is_ in the conda package.  The install error seems to indicate that the install process is looking for the `build_log.json` file in a particular location that looks to me like an installed location, which, as the package has yet to be installed, makes no sense (to me).

Thanks for your patience with all of this, @analog-cbarber.  I really would like to solve this problem

**Reply by analog-cbarber** [reply]
> > It [`whl2conda install`] is just doing a conda install of the package but since conda does not install dependencies in that case we read the dependencies and install them separately also using `conda install`.
> 
> * How do I "read the dependencies?"

You don't need to because `whl2conda install` does that, but the dependencies are listed in the `info/index.json` file within the conda package. You can read about the conda package format here: https://docs.conda.io/projects/conda-build/en/stable/resources/package-spec.html

And the [whl2conda reference documentation](https://zuzukin.github.io/whl2conda/latest/reference/links.html) lists other relevant specs if you are interested.

> * Also, the `whl2conda` documentation indicates that `whl2conda install` installs the package _with dependencies_.

That is correct. I was just explaining what `whl2conda install` is doing.

> * Would `conda install` really not install dependencies?

That is right. See https://github.com/conda/conda/issues/13043

> > One thing to try is to just do `conda install -p <conda-env-path> <conda-package-file>` and see if you get the same
> > error.
> 
> I tried the above `conda install` as you suggested, but I do get the same error.
> 
> > You can also build your conda package using the `tree` output format ...
> 
> Build the package with the "`tree` output format" of what command?

That is an `--format` option to `whl2conda convert`. If you give it `--format tree`, then instead of generating a `.conda` file or a `.tar.bz2` file, a directory will be created with the unpacked conda package contents in it. You can also just generate the `.conda` file and then use the `cph` tool (from the `conda-package-handling` package, which is a dependency of whl2conda) extract the directory tree:

```bash
$ cph extract mypackage.conda
```

The only reason to use `--format tree` is either because you want to debug what whl2conda is doing or you want to modify the package contents before packaging it up using `cph create`.

In any case, I don't think I have enough information to determine what is going on here. Can you upload your
`info/` directory from your built conda package and also the `*.dist-info/` subdirectory from your unpacked wheel
(use `wheel unpack` to get that)?

Another thing you can try would be to try building and installing on a non-Windows machine (assuming you have one and the code is not Windows specific).

**Reply by celticflute** [reply]
> Can you upload your `info/` directory from your built conda package and also the `*.dist-info/` subdirectory from your unpacked wheel

@analog-cbarber:  Please find attached the `info/` and `*dist-info/` directories you requested above.  Thank you so much for your continued assistance in this matter.
[info.zip](https://github.com/user-attachments/files/15955949/info.zip)
[jupyterlab_hide_code-4.0.0.dist-info.zip](https://github.com/user-attachments/files/15955956/jupyterlab_hide_code-4.0.0.dist-info.zip)
